### PR TITLE
feat(api): V1 API Phase 3 - Client tools, tool results, component streaming

### PR DESCRIPTION
**Pull request by alecf**
## Summary

Completes **V1 API Phase 3** as defined in `plans/implement-api-v1.md`. This phase adds the streaming and tool handling capabilities to the V1 API.

### Phase 3 Tasks Completed

Per the plan, Phase 3 covers: _"SSE streaming endpoints, tool call handling, atomic concurrency control."_

#### 3.1 Atomic Run Start ✅
- `startRun()` in `v1.service.ts` uses atomic conditional UPDATE to prevent race conditions
- Clear last run outcome fields when starting new run
- Returns proper 409 Conflict for concurrent runs

#### 3.2 Run Endpoints ✅ (completed in PR #1792)
- `POST /v1/threads/runs` - Create thread with run (SSE)
- `POST /v1/threads/:threadId/runs` - Create run on existing thread (SSE)
- `DELETE /v1/threads/:threadId/runs/:runId` - Cancel run
- SSE headers and connection close handling

#### 3.3 SSE Event Emission ✅ (completed in PR #1792)
- Simple `emit()` helper for AG-UI events
- `executeRun()` streams RUN_STARTED, content events, RUN_FINISHED
- Error handling emits RUN_ERROR with sanitized messages

#### 3.4 Tool Call Handling ✅ (this PR)
- **Client-side tools**: Detected via `ToolCallTracker` when tool calls don't receive results
- **`awaiting_input` event**: Emitted when client tools are pending, includes tool call IDs and arguments
- **Tool result processing**: `extractToolResults()`, `validateToolResults()`, `convertToolResultsToMessages()`
- **Continuation flow**: `previousRunId` validation, tool result messages saved before run starts

#### 3.5 Component Streaming ✅ (this PR)
- **`ComponentStreamTracker`**: Tracks incremental JSON parsing with `partial-json` library
- **`tambo.component.start`**: Emitted on first delta
- **`tambo.component.props_delta`**: RFC 6902 JSON Patch operations with streaming status per property
- **`tambo.component.end`**: Emitted with final props on tool call completion
- **Integration**: Hooked into `ai-sdk-client.ts` for component tools (`show_component_*`)

### Files Created

| File | Purpose |
|------|---------|
| `apps/api/src/v1/v1-tool-conversions.ts` | Convert V1 JSON Schema tools/components to internal format |
| `apps/api/src/v1/v1-tool-conversions.test.ts` | 21 tests for tool/component conversion |
| `apps/api/src/v1/v1-client-tools.ts` | `ToolCallTracker` class and `createAwaitingInputEvent()` |
| `apps/api/src/v1/v1-client-tools.test.ts` | 17 tests for tool call tracking |
| `apps/api/src/v1/v1-tool-results.ts` | Tool result extraction, validation, conversion |
| `apps/api/src/v1/v1-tool-results.test.ts` | 9 tests for tool result handling |
| `packages/backend/src/util/component-streaming.ts` | `ComponentStreamTracker` for props streaming |
| `packages/backend/src/util/component-streaming.test.ts` | 17 tests for component streaming |

### Files Modified

| File | Changes |
|------|---------|
| `apps/api/src/v1/v1.service.ts` | Integrated tool conversion, client tool tracking, tool result validation, awaiting_input emission |
| `apps/api/src/v1/v1.errors.ts` | Added `INVALID_TOOL_RESULT` error type |
| `packages/backend/src/services/llm/ai-sdk-client.ts` | Component streaming integration in `handleStreamingResponse()` |
| `packages/db/src/operations/thread.ts` | `clearPendingToolCalls()` with validation |

### Key Implementation Details

**Tool Call Tracking Flow:**
```
LLM generates tool call
  ↓
ToolCallTracker.processEvent() tracks TOOL_CALL_START/ARGS/END
  ↓
If no TOOL_CALL_RESULT received → tool is client-side
  ↓
Emit tambo.run.awaiting_input with pendingToolCallIds
  ↓
Save pendingToolCallIds to thread, set lastCompletedRunId
  ↓
Client submits tool_result content blocks with previousRunId
  ↓
validateToolResults() ensures all pending IDs have results
  ↓
Convert to internal messages, save, continue run
```

**Component Streaming Flow:**
```
tool-input-start for show_component_*
  ↓
Create ComponentStreamTracker
  ↓
tool-input-delta → processJsonDelta()
  ↓
Parse with partial-json, detect new/changed props
  ↓
Emit tambo.component.start (first delta)
Emit tambo.component.props_delta (JSON Patch + streamingStatus)
  ↓
tool-call complete → finalize()
  ↓
Emit tambo.component.end with finalProps
```

**Fail-Fast Patterns (per AGENTS.md):**
- 10MB limit for component JSON - throws on exceed
- 1MB limit for tool arguments - throws on exceed
- Unparseable final JSON in component - throws with context
- Missing/extra tool results - returns 400 with details
- Invalid previousRunId - returns 400 with expected vs actual

### Review Feedback Addressed

**P1 (Critical):**
- Tool result conversion preserves `toolUseId` as `tool_call_id`
- Transaction handling saves tool result messages within run transaction

**P2 (Important):**
- Size limits with fail-fast error handling
- Efficient string accumulation with array+join
- Property streaming status fix (only "done" when NEW property appears)
- Use `fast-deep-equal` instead of custom implementation
- DB validation with `.returning()` on clearPendingToolCalls

## Test plan

- [x] 47 new API tests pass (v1-client-tools, v1-tool-conversions, v1-tool-results)
- [x] 17 new component-streaming tests pass
- [x] All existing tests pass
- [x] `npm run check-types` passes
- [x] `npm run lint` passes
- [x] Size limit enforcement tests
- [x] Edge case: TOOL_CALL_ARGS without prior start (graceful handling)
- [x] Edge case: unparseable JSON in finalize() (throws with context)

## What's Next (Phase 4)

Per the plan, Phase 4 covers: _"Component state updates, comprehensive integration tests."_
- Component state endpoint: `POST /v1/threads/:threadId/components/:componentId/state`
- JSON Patch support for state updates
- E2E integration tests for full flows

🤖 Generated with [Claude Code](https://claude.com/claude-code)

### CODE REVIEW

File: apps/api/src/v1/v1-client-tools.ts [hunk 1]
```diff
@@ -0,0 +1,159 @@
...
+        }
+        break;
+      }
+      case EventType.TOOL_CALL_RESULT: {
+        // Tool call has been handled (either system tool or UI tool)
+        const e = event as unknown as ToolCallResultEvent;
+        this.completedToolCallIds.add(e.toolCallId);
+        break;
+      }
+    }
```
[the remark this answers is not in the document]
Comment (alecf) [reply apps/api/src/v1/v1-client-tools.ts]: @CharlieHelps yes please

File: apps/api/src/v1/v1-tool-results.ts [hunk 2]
```diff
@@ -0,0 +1,172 @@
...
+      error: {
+        code: "EXTRA_RESULTS",
+        message: `Tool results provided for unknown tool calls: ${extraToolCallIds.join(", ")}`,
+        extraToolCallIds,
+      },
+    };
+  }
+
+  return { valid: true };
+}
```
[the remark this answers is not in the document]
Comment (alecf) [reply apps/api/src/v1/v1-tool-results.ts:148]: @CharlieHelps yes please - eliminate duplicates, and warn about it (it's not a critical error, but the warning would be informative if it ever occurs)

File: apps/api/src/v1/v1.service.ts [hunk 3]
```diff
@@ -379,12 +391,53 @@ export class V1Service {
...
+            ),
+            HttpStatus.BAD_REQUEST,
+          ),
+        };
+      }
+
+      // Convert validated tool results to internal messages
+      toolResultMessages.push(...convertToolResultsToMessages(toolResults));
+    }
+
```
[the remark this answers is not in the document]
Comment (alecf) [reply apps/api/src/v1/v1.service.ts]: @CharlieHelps yes please

File: packages/backend/src/util/component-streaming.ts [hunk 4]
```diff
@@ -0,0 +1,266 @@
...
+
+/**
+ * Streaming status for component properties.
+ */
+export type PropStreamingStatus = "started" | "streaming" | "done";
+
+/**
+ * RFC 6902 JSON Patch operation.
+ */
+export interface JsonPatchOperation {
```
Comment (alecf) [inline packages/backend/src/util/component-streaming.ts]: @CharlieHelps there should be a json patch library that we can get json patch types from? (we're going to need it anyway in the react-sdk at some point in order to *apply* patches)

File: packages/backend/src/util/component-streaming.ts [hunk 5]
```diff
@@ -0,0 +1,266 @@
...
+    }
+
+    // Emit props_delta event if there are changes
+    if (patches.length > 0) {
+      events.push(this.createPropsDeltaEvent(patches));
+    }
+
+    // Update previous props (use JSON parse/stringify for simple deep clone)
+    this.previousProps = JSON.parse(JSON.stringify(currentProps));
+
```
[the remark this answers is not in the document]
Comment (alecf) [reply packages/backend/src/util/component-streaming.ts:165]: @CharlieHelps yes please (seems like structuredClone is best)

File: packages/backend/src/util/component-streaming.ts [hunk 6]
```diff
@@ -0,0 +1,266 @@
...
+    // Mark all properties as done
+    for (const key of Object.keys(this.streamingStatus)) {
+      this.streamingStatus[key] = "done";
+    }
+
+    // Emit end event
+    events.push(this.createEndEvent(finalProps));
+
+    return events;
+  }
```
[the remark this answers is not in the document]
Comment (alecf) [reply packages/backend/src/util/component-streaming.ts:204]: @CharlieHelps yes please

File: packages/backend/src/util/component-streaming.ts [hunk 7]
```diff
@@ -0,0 +1,266 @@
...
+
+    // Try to parse incrementally
+    let currentProps: Record<string, unknown>;
+    try {
+      const accumulatedJson = this.jsonChunks.join("");
+      currentProps = parse(accumulatedJson) as Record<string, unknown>;
+    } catch {
+      // Can't parse yet, wait for more data
+      return events;
+    }
```
[the remark this answers is not in the document]
Comment (alecf) [reply packages/backend/src/util/component-streaming.ts]: @CharlieHelps yes please

File: packages/backend/src/services/llm/ai-sdk-client.ts [hunk 8]
```diff
@@ -467,10 +475,30 @@ export class AISdkClient implements LLMClient {
...
+              componentName,
+            );
+          } else {
+            componentTracker = undefined;
+          }
           break;
         case "tool-input-delta":
           accumulatedToolCall.arguments += delta.delta;
           toolCallArgDeltas = [...toolCallArgDeltas, delta.delta];
+
```
Comment (MichaelMilstead) [inline packages/backend/src/services/llm/ai-sdk-client.ts:497]: For some reason when starting a run that results in showing a component, the response first streams all the custom component events (which is correct I guess):
`data: {"type":"CUSTOM","name":"tambo.component.props_delta","value":{"componentId":"message-HlPzK4GpQ9WVjVzSwLRXR","patch":[{"op":"add","path":"/location","value":""}],"streamingStatus":{"_tambo_displayMessage":"done","_tambo_statusMessage":"done","_tambo_completionStatusMessage":"done","location":"started"}},"timestamp":1768932969123}`

But then once those are all complete, it streams all the "tool call" events for the same thing:
```
data: {"type":"TOOL_CALL_START","toolCallId":"call_3CnKmxE85sVmXyuOSGQuGUop","toolCallName":"show_component_WeatherCard","timestamp":1768932969587}

data: {"type":"TOOL_CALL_ARGS","toolCallId":"call_3CnKmxE85sVmXyuOSGQuGUop","delta":"{\"","timestamp":1768932969587}

...
```

Followed by the `TEXT_MESSAGE_CONTENT` events, which seems correct.

I guess we need to make sure when we emit the custom component events we don't also emit tool call events for showing the component so the client doesn't try to call a tool that doesn't exist?

Also, I think we could have it not include those `_tambo` fields for the component events just to reduce chunks

File: apps/api/src/v1/v1-client-tools.ts [hunk 9]
```diff
@@ -0,0 +1,166 @@
...
+        this.toolCallArgumentSizes.delete(e.toolCallId);
+        break;
+      }
+      case EventType.TOOL_CALL_RESULT: {
+        const e = event as unknown as ToolCallResultEvent;
+        this.startedToolCalls.delete(e.toolCallId);
+        this.toolCallArgumentChunks.delete(e.toolCallId);
+        this.toolCallArgumentSizes.delete(e.toolCallId);
+        break;
+      }
```
[the remark this answers is not in the document]
Comment (alecf) [reply apps/api/src/v1/v1-client-tools.ts:147]: @CharlieHelps yes please - I'm not actually sure if the ToolCallTracker is tracking just client tools, or all tools? perhaps we just need explicit accounting to distinguish between client and server tools?

File: apps/api/src/v1/v1-client-tools.ts [hunk 10]
```diff
@@ -0,0 +1,166 @@
...
+    value: {
+      pendingToolCalls: pendingToolCalls.map((tc) => ({
+        toolCallId: tc.toolCallId,
+        toolName: tc.toolName,
+        arguments: tc.arguments,
+      })),
+    },
+    timestamp: Date.now(),
+  } as BaseEvent;
+}
```
[the remark this answers is not in the document]
Comment (alecf) [reply apps/api/src/v1/v1-client-tools.ts:211]: @CharlieHelps yes please - I think there is a type for CustomEvent that we can use, something like `type AwaitingInputEvent = CustomEvent & {...}`;

File: apps/api/src/v1/v1-tool-conversions.ts [hunk 11]
```diff
@@ -0,0 +1,161 @@
...
+  const result = new AvailableComponentDto();
+  result.name = component.name;
+  result.description = component.description;
+  // V1 doesn't have component-specific contextTools, they're passed separately
+  result.contextTools = [];
+  // Store the propsSchema - the internal system will use it for component generation
+  // The props field stores metadata about the component's props
+  result.props = component.propsSchema as unknown as typeof result.props;
+  return result;
+}
```
[the remark this answers is not in the document]
Comment (alecf) [reply apps/api/src/v1/v1-tool-conversions.ts:169]: @CharlieHelps yes please

File: packages/backend/src/services/llm/ai-sdk-client.ts (hunk 8 above)
Comment (alecf) [reply packages/backend/src/services/llm/ai-sdk-client.ts:497]: Hmmm. I'm honestly not sure if we should emit both or not! I'd like to clean that up during the SDK phase - I think we'll find this and other subtle bugs as we try to consume the stream, and it may be more obvious what the proper behavior is (i.e. if I would have to write a lot of custom code to work around the extra events, that seems bad... or alternatively the extra events might help us keep track of which "tools" are done or not... not sure)

File: packages/backend/src/util/component-streaming.ts [hunk 12]
```diff
@@ -0,0 +1,264 @@
...
+export function isComponentTool(toolName: string): boolean {
+  return toolName.startsWith("show_component_");
+}
+
+/**
+ * Extract component name from tool name.
+ */
+export function extractComponentName(toolName: string): string {
+  return toolName.replace("show_component_", "");
+}
```
[the remark this answers is not in the document]
Comment (alecf) [reply packages/backend/src/util/component-streaming.ts:353]: @CharlieHelps yes please

File: packages/backend/src/util/component-streaming.ts [hunk 13]
```diff
@@ -0,0 +1,297 @@
...
+
+    // Emit props_delta event if there are changes
+    if (patches.length > 0) {
+      events.push(this.createPropsDeltaEvent(patches));
+    }
+
+    this.previousProps = structuredClone(currentProps);
+
+    return events;
+  }
```
[the remark this answers is not in the document]
Comment (alecf) [reply packages/backend/src/util/component-streaming.ts:167]: @CharlieHelps yes please - make sure to document this well so we don't try to do structureClone or other expensive operations

File: packages/backend/src/util/component-streaming.ts [hunk 14]
```diff
@@ -0,0 +1,297 @@
...
+    // Check for removed properties (unlikely in streaming but handle it)
+    for (const key of Object.keys(previousProps)) {
+      if (!(key in currentProps)) {
+        patches.push({ op: "remove", path: `/${key}` });
+        delete this.streamingStatus[key];
+      }
+    }
+
+    return { patches, statusUpdates, newPropertyKeys };
+  }
```
[the remark this answers is not in the document]
Comment (alecf) [reply packages/backend/src/util/component-streaming.ts:260]: @CharlieHelps yes please - do any of the json-patch libraries that we already  use have an escaper that we can use instead of rolling our own? I don't mind rolling our own if not

File: packages/backend/src/util/component-streaming.ts [hunk 15]
```diff
@@ -0,0 +1,316 @@
...
+const COMPONENT_TOOL_PREFIX = "show_component_";
+
+/**
+ * Streaming status for component properties.
+ */
+export type PropStreamingStatus = "started" | "streaming" | "done";
+
+export type JsonPatchOperation = Operation;
+
+/**
```
[the remark this answers is not in the document]
Comment (alecf) [reply packages/backend/src/util/component-streaming.ts:36]: @CharlieHelps as per agent directions we should not be re-exporting types anyway, especially from 3rd party libraries - can you just update consumers to use the fast-json-patch types directly?

File: packages/db/src/operations/thread.ts [hunk 16]
```diff
@@ -422,6 +422,34 @@ export async function updateThreadGenerationStatus(
...
+    })
+    .where(eq(schema.threads.id, threadId))
+    .returning({ id: schema.threads.id });
+
+  if (!updated) {
+    throw new Error(
+      `Failed to clear pending tool calls: Thread ${threadId} not found`,
+    );
+  }
+}
```
[the remark this answers is not in the document]
Comment (alecf) [reply packages/db/src/operations/thread.ts:501]: @CharlieHelps yes please

File: packages/db/src/operations/thread.ts [hunk 17]
```diff
@@ -422,6 +422,56 @@ export async function updateThreadGenerationStatus(
...
+      pendingToolCallIds: null,
+      updatedAt: sql`now()`,
+    })
+    .where(and(...whereConditions))
+    .returning({ id: schema.threads.id });
+
+  if (!updated) {
+    throw new PendingToolCallStateMismatchError(threadId);
+  }
+}
```
[the remark this answers is not in the document]
Comment (alecf) [reply packages/db/src/operations/thread.ts:501]: @CharlieHelps yes please

**Review by MichaelMilstead — APPROVED** [review]
